### ISSUE TITLE
Segfaulting on matrix multiplication with MKL

### ISSUE THREAD

**Issue by paulkass** [raskr/rust-autograd#22]
Matrix multiplication can cause a segault when using MKR. Here is a snippet:

```
let W = ag::variable(ag::ndarray_ext::glorot_uniform(&[(output_size as usize)*4, input_size as usize]));
let s = ag::slice(&W, &[0, 0], &[output_size, -1]);
let r = ag::matmul(&s, &x);
```

In the above example, `x` has dimensions `[input_size, 1]` and `s` will have dimensions `[output_size, input_size]`, so matrix multiplication works out. The example runs with no issues without the `mkr` feature, but with `mkr` the eval step segfaults. I analyzed with valgrind, and it seems like there a few extra writes happening somewhere, resulting in a segfault.

**Comment by raskr** [comment]
Probably fix in v0.9.6.
That was a bug of MatMul operation using Intel MKL. As you reported, I observed UB when an input to MatMul is a vertical vector. https://github.com/raskr/rust-autograd/blob/master/tests/test_tensor_ops_eval.rs#L54

**Comment by paulkass** [comment]
It seems like this an issue affecting mostly large matrices done with MKL. Without MKL, the matrix computation works fine but it's super slow. With MKL, the code segfaults. Here is a gist with the code:

https://gist.github.com/paulkass/2a94aa07d1ba43f8efbf7256d7aa2d3c

At sizes 1-10, it seems fine. 
At sizes 11-100, the multiplication is performed but the program still segfaults.
At sizes 100+, the program segfaults without any other output.

Hopefully this helps.

**Comment by raskr** [comment]
Ahhh yes, observed segv with your setting...
I'll check it later.

**Comment by raskr** [comment]
Fix in v0.9.7!
Sorry for the inconvenience...

**Comment by paulkass** [comment]
awesome, thanks!
